### ISSUE TITLE
ElementConfig should be used to determine props required

### ISSUE THREAD

**Issue by callumlocke** [yannickcr/eslint-plugin-react#1904]
ESLint currently uses `ElementProps` to determine what types are required by a component with Flow-annotated props, but it should be using `ElementConfig` instead. Flow docs [here](https://flow.org/en/docs/react/types/#toc-react-elementprops) back this up.

@ljharb [suggested](https://github.com/airbnb/javascript/issues/1843#issuecomment-401390862) I open an issue here to get it changed. Summary: perhaps unintuitively, the set of attributes explicitly set in a JSX expression does *not* correspond exactly with the props object that is passed into a component. The latter is always automatically augmented with any `defaultProps` *before* being passed in (this becomes obvious when you look at what a JSX expression compiles to). Therefore any prop with a corresponding defined key in `defaultProps` is _always_ defined in the props object received by the function, whether or not it was explicitly set with a JSX attribute. So it is wrong for eslint-plugin-react to require a question-mark on type annotations for props that have a corresponding key in defaultProps, as this annotation is only for the augmented object being passed into the function.

**Comment by okovpashko** [comment]
Totally agree.

Stumbled upon this issue when switched to `eslint-config-airbnb` and got a lot of Eslint errors in code that's written according to the [Flow docs](https://flow.org/en/docs/react/components/#toc-using-default-props).

Flow allows using this component without passing `name` prop, but Eslit shows an error, that there shouldn't be default prop defined for the required prop.
But if we make the `name` prop optional, we'll get Flow error that we can't call method `toUpperCase` of undefined.

Looks like an incompatibility between Eslint and Flow.

```javascript
type Props = {
	// propType "name" is required and should not have a defaultProp declaration react/require-default-props
	name: string,
};

class Greeting extends React.Component<Props> {
	static defaultProps = {
		name: 'User',
	};

	render () {
		// here will be Flow error if the "name" prop is optional
		// Cannot call `name.toUpperCase` because property `toUpperCase` is missing in  undefined [1].
		return <span>Hello, {this.props.name.toUpperCase()}</span>
	}
}
```

**Comment by jdeff** [comment]
I am running into the same issue when using TypeScript, FWIW.
